### ISSUE TITLE
Vector of types containing references fails to compile

### ISSUE THREAD

**Issue by DiligentGraphics** [electronicarts/EASTL#264]
The following code fails to compile:

```
struct MyStruct{};
vector<pair<const MyStruct&, int>> vec;
```

with the following output (MSVC):

```
> eastl-test\EASTL\include\EASTL/utility.h(402): error C2535: 'eastl::pair<const main::MyStruct &,int>::pair(T1)': member function already defined or declared
1>        with
1>        [
1>            T1=const main::MyStruct &
1>        ]
1>eastl-test\EASTL\include\EASTL/utility.h(390): note: see declaration of 'eastl::pair<const main::MyStruct &,int>::pair'
1>eastl-test\EASTL\include\EASTL/vector.h(431): note: see reference to class template instantiation 'eastl::pair<const main::MyStruct &,int>' being compiled
1>eastl-test\EASTL\include\EASTL/vector.h(429): note: while compiling class template member function 'eastl::VectorBase<T,Allocator>::~VectorBase(void)'
1>        with
1>        [
1>            T=eastl::pair<const main::MyStruct &,int>,
1>            Allocator=eastl::allocator
1>        ]
1>eastl-test\EASTL\include\EASTL/vector.h(506): note: see reference to function template instantiation 'eastl::VectorBase<T,Allocator>::~VectorBase(void)' being compiled
1>        with
1>        [
1>            T=eastl::pair<const main::MyStruct &,int>,
1>            Allocator=eastl::allocator
1>        ]
1>eastl-test\EASTL\include\EASTL/vector.h(506): note: while compiling class template member function 'eastl::vector<eastl::pair<const main::MyStruct &,int>,eastl::allocator>::vector(void) noexcept(false)'
1>D:\GitHub\eastl-test\main.cpp(21): note: see reference to function template instantiation 'eastl::vector<eastl::pair<const main::MyStruct &,int>,eastl::allocator>::vector(void) noexcept(false)' being compiled
1>eastl-test\main.cpp(21): note: see reference to class template instantiation 'eastl::vector<eastl::pair<const main::MyStruct &,int>,eastl::allocator>' being compiled
1>eastl-test\EASTL\include\EASTL/utility.h(437): error C2535: 'eastl::pair<const main::MyStruct &,int>::pair(eastl::pair_first_construct_t,T1)': member function already defined or declared
1>        with
1>        [
1>            T1=const main::MyStruct &
1>        ]
1>eastl-test\EASTL\include\EASTL/utility.h(425): note: see declaration of 'eastl::pair<const main::MyStruct &,int>::pair'
```

It works fine on [gcc](https://godbolt.org/z/6ZG0C7), clang and msvc.

**Comment by rparolin** [comment]
I'm not convinced this is a legitimate issue given the standard specification.  std containers do not support holding references because they aren't re-assignable.  This is why the reference_wrapper type was introduced.  Can you point to standard citation that  speaks to this feature?

https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper

```std::reference_wrapper is a class template that wraps a reference in a copyable, assignable object. It is frequently used as a mechanism to store references inside standard containers (like std::vector) which cannot normally hold references.```

**Comment by rparolin** [comment]
Closing the issue but we can continue discussion.  If required, I can reopen the issue.

**Comment by DiligentGraphics** [comment]
I am not a standard expert, but all std containers can definitely contain types that are not assignable, but movable. In the example above, pair<const MyStruct&, int> is movable but not assignable type, so I guess it should be OK to hold it in a vector.
The code above compiles fine by all major compilers, so I am pretty convinced it is legitimate.

It is up to you if you want to fix this as we decided not to use EASTL for now: there is a lot of compiler errors especially for not copyiable types and unfortunately we did not see significant performance improvements vs std containers.
